azkaban-common/src/test/java/azkaban/jobExecutor/ProcessJobTest.java

```diff
@@ -136,4 +142,55 @@ public void testPartitionCommand() throws Exception {
     Assert.assertArrayEquals(new String[]{"a", "e=b c"},
         ProcessJob.partitionCommandLine(test3));
   }
+
+  /**
+   * test cancellation of the job before associated process is constructed
+   * expect job will be cancelled successfully
+   */
+  @Test
+  public void testCancelDuringPreparation() throws InterruptedException, ExecutionException {
+    final Props jobProps = new Props();
+    jobProps.put("command", "echo hello");
+    jobProps.put("working.dir", "/tmp");
+    jobProps.put("user.to.proxy", "test");
+    jobProps.put("azkaban.flow.projectname", "test");
+    jobProps.put("azkaban.flow.flowid", "test");
+    jobProps.put("azkaban.job.id", "test");
+    jobProps.put("azkaban.flow.execid", "1");
+
+    final Props sysProps = new Props();
+    sysProps.put("execute.as.user", "false");
+    final SleepBeforeRunJob job = new SleepBeforeRunJob("test", sysProps, jobProps, this.log);
+
+    final ExecutorService executorService = Executors.newSingleThreadExecutor();
+    final Future future = executorService.submit(job);
+    Thread.sleep(2000);
+    try {
+      this.job.cancel();
+    } catch (final Exception e) {
+      this.log.error(e.getMessage());
+    }
+    future.get();
+    assertEquals(this.job.getProgress(), 0.0, 0.001);
+  }
+
+  class SleepBeforeRunJob extends ProcessJob implements Runnable {
+
+    public SleepBeforeRunJob(final String jobId, final Props sysProps, final Props jobProps,
+        final Logger log) {
+      super(jobId, sysProps, jobProps, log);
+    }
+
+    @Override
+    public void run() {
+      try {
+        info("sleep for 5 seconds before actually running the job");
+        Thread.sleep(5 * 1000);
```

Comment:
this is basically to simulate the scenario where job preparation is stuck. 200ms/1000ms is too short for this purpose but i think 3secs is a reasonable number. let me change it to 3secs.